powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationServiceBehavior.java

```diff
@@ -716,6 +734,102 @@ private Map<String, String> mapMerge(Map<String, String> m1, Map<String, String>
         return m3;
     }
 
+    @SneakyThrows(GenericServiceException.class)
+    private void generateAndSetOtpToOperationDetail(final OperationEntity operation, final OperationDetailResponse operationDetailResponse) {
+        final String totp = generateTotp(operation, powerAuthServiceConfiguration.getProximityCheckOtpLength());
+        operationDetailResponse.setProximityOtp(totp);
+    }
+
+    private String generateTotp(final OperationEntity operation, final int otpLength) throws GenericServiceException {
+        final String seed = operation.getTotpSeed();
+        final String operationId = operation.getId();
+
+        if (seed == null) {
+            logger.debug("Seed is null for operation ID: {}", operationId);
+            return null;
+        }
+
+        try {
+            byte[] seedBytes = Base64.getDecoder().decode(seed);
+            final LocalDateTime now = LocalDateTime.now();
+            byte[] totp = Totp.generateTotpSha256(seedBytes, now, otpLength);
+
+            final AuditDetail auditDetail = createProximityOtpAuditDetail(operation, seed, now);
+            audit.log(AuditLevel.INFO, "Proximity OTP generated for operation ID: {}", auditDetail, operationId);
+
+            return new String(totp);
+        } catch (CryptoProviderException | IllegalArgumentException e) {
+            logger.error("Unable to generate OTP for operation ID: {}, user ID: {}", operationId, operation.getUserId(), e);
+            throw new GenericServiceException(ServiceError.OPERATION_ERROR, e.getMessage(), e.getLocalizedMessage());
+        }
+    }
+
+    private static String generateTotpSeed(final OperationCreateRequest request, final OperationTemplateEntity template) throws GenericServiceException {
```

Comment:
Opinion: I think logging inside this method is overly excessive - do we really need this much information and add two parameters to the method that can work with a simple boolean?